plot_anomaly_comparison-3D_SKlearn.py

```diff
@@ -0,0 +1,146 @@
+zline = (t_hex-(max(t_hex)+min(t_hex))/2)/(max(t_hex)-min(t_hex))*2 + sd * np.random.randn(n_inliers)
+X_hex=np.c_[xline,yline,zline]
+## 3: Sphere, equally distribution
+def fibonacci_sphere(samples=1,randomize=True):
+    rnd = 1.
+    if randomize:
+        rnd = np.random.random() * samples
+    points = []
+    offset = 2./samples
+    increment = np.pi * (3. - np.sqrt(5.));
+    for i in range(samples):
+        y = ((i * offset) - 1) + (offset / 2);
+        r = np.sqrt(1 - pow(y,2))
+        phi = ((i + rnd) % samples) * increment
+        x = np.cos(phi) * r
+        z = np.sin(phi) * r
+        points.append([x,y,z])
+    return points
+X_sph=np.array(fibonacci_sphere(samples=n_inliers))
+## 4: Gaussian Mixture
+blobs_params = dict(random_state=0, n_samples=n_inliers, n_features=3)
+X_gau=make_blobs(centers=[[-0.7, -0.7,-0.7], [0, 0,0],
+                          [0.7,0.7,0.7]], cluster_std=[0.2, 0.2,0.2],**blobs_params)[0]
+## 5: Misaligned Gaussian Mixture
+blobs_params = dict(random_state=0, n_samples=n_inliers, n_features=3)
+X_misaligned=make_blobs(centers=[[-0.7, -0.7,-0.7], [0.7, 0.7,-0.7],
+                          [-0.7, 0.7,0.7]], cluster_std=[0.2, 0.2,0.2],**blobs_params)[0]
+## 6: Whole dataset
+datasets3D = [X_lin,X_hex,X_sph,X_gau,X_misaligned]
+
+# Define to data label
+labels = np.concatenate([np.ones(n_inliers),-np.ones(n_outliers)], axis=0)  
+# lbel 1 as inliers, -1 as outliers
+
+# define outlier/anomaly detection methods to be compared
+anomaly_algorithms = [
+    ("Robust covariance", EllipticEnvelope(contamination=outliers_fraction)),
+    ("One-Class SVM", svm.OneClassSVM(nu=outliers_fraction, kernel="rbf",
+                                      gamma=0.1)),
+    ("Isolation Forest (IF)", IsolationForest(n_estimators=500,behaviour='new',
+                                         contamination=outliers_fraction,
+                                         random_state=42)),
+    ("Local Outlier Factor", LocalOutlierFactor(
+        n_neighbors=35, contamination=outliers_fraction))] 
+
+plt.figure(figsize=(14,15))
+plt.subplots_adjust(left=.02, right=.98, bottom=.001, top=.98, wspace=.05,hspace=.01)
+plot_num = 1
+rng = np.random.RandomState(42)
+noise_params = dict(random_state=0, n_samples=n_samples, n_features=3)
+AUC=np.array([])
+for i_dataset3D, X in enumerate(datasets3D):
+    # Add outliers (i = index (0,1,..), X = data set)
+    X = np.concatenate([X, rng.uniform(low=-1.2, high=1.2,
+                       size=(n_outliers, 3))], axis=0)                  # add uniform outlier
+    X_noise =rng.uniform(low=-1.2, high=1.2,size=(n_samples, D_noise))  # add uniform noise dimension
+    X = np.append(X, X_noise, axis=1)
+    AUC_row=np.array([])
+    for name, algorithm in anomaly_algorithms: 
+        t0 = time.time()
+        algorithm.fit(X)
+        t1 = time.time()
+
+        # fit the data and tag outliers
+        if name == "Local Outlier Factor":
+            y_pred = algorithm.fit_predict(X)
```

Comment:
I'm confused do the others not implement `fit_predict`, or would `fit().predict()` not work here?